### PR TITLE
Robustness detections in giskard.scan

### PR DESCRIPTION
**Pull request by mathieuroques**
I've added a text perturbation scan:

- a mask for locations, ethnicity, gender and proper noun
- add 20% of typos
- add 10% of random character modification
- turn into title case
- delete or replace by a random word 5% of the words

### CODE REVIEW

File: python-client/giskard/scanner/robustness/unit_perturbation.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+class TransformationGenerator:
+    def __init__(self, model, dataset):
+        self.model = model
+        self.dataset = dataset
+        self.column_types = dataset.column_types
+        self.text_perturbation_generator = TextTransformer()
+
+    def generate_std_transformation(self, feature):
+        mad = self.dataset.df[feature].mad()
```
Comment (mattbit) [inline python-client/giskard/scanner/robustness/unit_perturbation.py:14]: pandas `mad` is deprecated and it is not what we want. We want "median absolute deviation" which is a robust estimator of the standard deviation, and not the "mean absolute deviation".

σ ≈ 1.4826 * MAD

where MAD is `median(abs(x - median(x)))`.

**Comment by mathieuroques** [comment]
- Still have to add the religious key-words 
 - 💡 translate keywords with https://huggingface.co/t5-base

File: python-client/giskard/scanner/robustness/unit_perturbation.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+                return x
+
+        return func
+
+
+class TextTransformer:
+    def __init__(self):
+        self.text = None
+        self.tokenizer = AutoTokenizer.from_pretrained("dslim/bert-base-NER")
+        self.model = AutoModelForTokenClassification.from_pretrained("dslim/bert-base-NER")
```
Comment (mattbit) [inline python-client/giskard/scanner/robustness/unit_perturbation.py:43]: This is an English language model

**Review by mattbit — CHANGES_REQUESTED** [review]
Great work! We should split the different transformations to apply them separately.
We also need to think about support for multiple languages (at least French and English to begin).

File: python-client/giskard/scanner/robustness/unit_perturbation.py (hunk 2 above)
Comment (mathieuroques) [reply python-client/giskard/scanner/robustness/unit_perturbation.py:43]: I'll change it with https://huggingface.co/Davlan/distilbert-base-multilingual-cased-ner-hrl

**Comment by mattbit** [comment]
Refactored to simple text perturbations.

- need to add again the numerical perturbation
- need to add more text transformations

**Review by mattbit — APPROVED** [review]

File: python-client/giskard/slicing/text_slicer.py [hunk 3]
```diff
@@ -85,44 +80,103 @@ def find_metadata_slices(self, feature, target):
...
-        # Compute the global TF-IDF score for each word.
-        global_tfidf = np.asarray(tfidf.mean(axis=0)).squeeze()
-        sorted_global_tfidf_indices = global_tfidf.argsort()[::-1]
+        mask = (counts.max(axis=-1) > 5) & (counts.min(axis=-1) > 0) & (remainders.min(axis=-1) > 0)
 
-        # Get the top n words sorted by global TF-IDF score.
-        top_words = [inv_vocab[idx] for idx in sorted_global_tfidf_indices[:n]]
+        _data = []
+        for token, token_counts, token_remainders in zip(tokens[mask], counts[mask], remainders[mask]):
+            print(token_counts, token_remainders)
```
Comment (andreybavt) [inline python-client/giskard/slicing/text_slicer.py:148]: is this print for debugging purposes or you actually wanted to present it to a user?

**Review by andreybavt — CHANGES_REQUESTED** [review]

File: python-client/giskard/scanner/logger.py [hunk 4]
```diff
@@ -0,0 +1,4 @@
+import logging
+
+
+logger = logging.getLogger(__name__)
```
[the remark this answers is not in the document]
Comment (mattbit) [reply python-client/giskard/scanner/logger.py:4]: Yep I wanted a logger which is common for the "scan" package.

**Comment by mattbit** [comment]
To be merged when linter approves!

**Pull request merged**